Add org.ultimatepp.TheIDE

klugier opened the pull request:
### Please confirm your submission meets all the criteria

- [x] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [x] My pull request follows the instructions at [App Submission][submission].
- [x] I have [built][build] and tested the submission locally.
    - I tested only on x86_64. The code and Flatpak is compatible with aarch64 too. However, I don't have such machine at home to verify.
- [x] I am using only the minimal set of permissions. *(If not, please explain each non-standard permission.)*
    - --socket=x11 - the application currently runs only on X11 via GTK 3 backend
    - --share=network - network is needed by application for example to execute HttpRequest 
    - --share=ipc - necessary for X11
    - --filesystem=host - access to host directory is needed. The compilation done by IDE is executed on host not within sandbox. So, broad access is needed.
    - --talk-name=org.freedesktop.Flatpak - required for host-spawn, essential part of this Flatpak
        -  https://github.com/flathub/flatpak-builder-lint/pull/227
- [x] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
- [x] I am an upstream contributor to the project. If not, I contacted upstream developers about submitting their software to Flathub. [U++ on GitHub - People](https://github.com/orgs/ultimatepp/people)
    - I am upstream contributor with owner rights to ultimatepp organization on GitHub
- [x] I own the domain used in the application ID or the domain has a policy for delegating subdomains (e.g. GitHub, SourceForge).
    - The owner of the ultimatepp.org is Miroslav Filder. However, he gave me all required permission to move forward with Flatpak publication.
- [x] Any additional patches or files have been submitted to the upstream projects concerned. *(If not, explain why.)*
    - Currently under review: https://github.com/ultimatepp/ultimatepp/pull/177. Will be merged once the Flatpak package is published, and the changes will be present in the new version 2024.1. For the old one targeted in this PR, a patch file is required.

[reqs]: https://docs.flathub.org/docs/for-app-authors/requirements
[maint]: https://docs.flathub.org/docs/for-app-authors/maintanance
[submission]: https://docs.flathub.org/docs/for-app-authors/submission
[build]: https://docs.flathub.org/docs/for-app-authors/submission/#before-submission

hfiguiere commented on org.ultimatepp.TheIDE.desktop:
```diff
@@ -0,0 +1,12 @@
+Comment=U++ IDE
+MimeType=application/x-upp;
+Exec=run-ide.sh
+Icon=org.ultimatepp.TheIDE.png
```
never an extension for the icon name.

hfiguiere commented on org.ultimatepp.TheIDE.yml:
```diff
@@ -0,0 +1,115 @@
+  - --socket=x11
+  - --share=network
+  - --share=ipc
+  - --filesystem=home
+  - --filesystem=host
```
these are very broad and redundant.

klugier replied:
I didn't now that host also covers home. However, after removing home it looks like the access to that directory works as expected. home removed in latest version.

--filesyste=host premision is needed since the application is IDE. Here is the quotation from PR description "access to host directory is needed. The compilation done by IDE is executed on host not within sandbox. So, broad access is needed.". I also notcied that other IDE's such as VS Code also using this permission.

Reference:
- https://github.com/flathub/com.visualstudio.code/blob/c2704e2c9c31ba12f8266fc27d5c73f6e68a51f4/com.visualstudio.code.yaml#L20

hfiguiere replied:
> I also notcied that other IDE's such as VS Code also using this permission.

two wrongs don't make a right.

as for "host" I don't see why you'd need anything out of the home directory.

klugier replied:
Without access to host filesystem. This Flatpak doesn't make sense. TheIDE It is whole compilation and language environment. You are compiling your program on host not within sandbox.

Another example (Code::Blocks) for C++. It uses both --filesystem=home and --filesystem=host. I don't understand why I can not use it here and they can. Do you have a policy about it?

References:
- https://github.com/flathub/org.codeblocks.codeblocks/blob/d660f090f2e9726d1476d73ac200c7b74f4b8867/org.codeblocks.codeblocks.yaml#L15

hfiguiere commented on org.ultimatepp.TheIDE.yml:
```diff
@@ -0,0 +1,115 @@
+  - --share=ipc
+  - --filesystem=home
+  - --filesystem=host
+  - --talk-name=org.freedesktop.Flatpak
```
this is not allowed by default.

klugier replied:
I just noticed that VS Code is using this permission. TheIDE is integrated development environment and without it, it simply won't work. TheIDE is exactly in the same category as VS Code is.

What is your recommendation here?

bbhtt replied:
https://docs.flathub.org/docs/for-app-authors/linter#exceptions

hfiguiere commented on org.ultimatepp.TheIDE.yml:
```diff
@@ -0,0 +1,115 @@
+    buildsystem: simple
+    build-commands: 
+      - printf "#define flagFlatpak 1\\n" > uppsrc/uppconfig.h
+      - git apply flatpak.patch
```
this is not how you add patches to a build.

klugier replied:
I just noticed that there is type: patch. I will rework it.

hfiguiere commented on org.ultimatepp.TheIDE.yml:
```diff
@@ -0,0 +1,115 @@
+        path: org.ultimatepp.TheIDE.desktop
+      - type: file
+        path: org.ultimatepp.TheIDE.metainfo.xml
+      - type: file
+        path: res/TheIDE_512x512.png
+      - type: file
+        path: res/TheIDE_256x256.png
+      - type: file
+        path: res/TheIDE_128x128.png
+      - type: file
+        path: res/TheIDE_64x64.png
```
only ONE icon since they don't come from upstream.

hfiguiere replied:
no icon upstream?

hfiguiere replied:
you can also submit a patc upstream for decently sized icon, and a SVG.

klugier replied:
Patch has been submitted to the upstream (https://github.com/ultimatepp/ultimatepp/pull/181) and accepted. Since, I can not use directly what is in the upstream due to lack of new release. I created simple patch from this PR (patch/icon.patch). Once the new version will be release it won't be needed.

hfiguiere commented on org.ultimatepp.TheIDE.yml:
```diff
@@ -0,0 +1,115 @@
+app-id: org.ultimatepp.TheIDE
+runtime: org.gnome.Platform
```
why do you need the gnome platform?

klugier replied:
TheIDE is using libnotify for notification. Without using org.gnome.Platform there is a compilation error.

Do you know better alternative? Maybe there is a way to add libnotify when your runtime is org.freedesktop.Platform? On the other hand on Flatpak documentation [page](https://docs.flatpak.org/en/latest/available-runtimes.html) dedicated to run times there is an information whenever you need libnotify GNOME platform should be used.

bbhtt replied:
libnotify will be in 24.08 of Freedesktop runtime, for now build it. It's small and doesn't have many dependencies.

bbhtt replied:
libnotify will be in freedesktop 24.08 runtime. If you need that only, you can build libnotify separately and remove it when bumping to 24.08

klugier replied:
Hello @bbhtt, Thanks for suggestion! I migrated to org.freedesktop.Platform and I added separate step to build libnotify from scratch.

hfiguiere commented on org.ultimatepp.TheIDE.yml:
```diff
@@ -0,0 +1,115 @@
+  - name: run-ide
+    buildsystem: simple
+    build-commands:
+      - install -D run-ide.sh /app/bin/run-ide.sh
```
```suggestion
      - install -Dm755 run-ide.sh /app/bin/run-ide.sh
```

hfiguiere commented on org.ultimatepp.TheIDE.yml:
```diff
@@ -0,0 +1,115 @@
+      - mv 3p/Terminal/PtyAgent app/PtyAgent
+      - mv 3p/Terminal/Terminal app/Terminal
+      - 3p/umk/umk app/,3p/uppsrc UppTerm 3p/umk/GCC.bm -brv +GUI,SHARED upp-term
+      - install -D upp-term /app/bin/upp-term
```
```suggestion
      - install -Dm755 upp-term /app/bin/upp-term
```

hfiguiere commented on org.ultimatepp.TheIDE.desktop:
Why is this file here when it exists upstream https://github.com/ultimatepp/ultimatepp/blob/master/uppsrc/ide/theide.desktop

klugier replied:
OK, I will rework it to use upstream desktop file.

hfiguiere commented on org.ultimatepp.TheIDE.yml:
```diff
@@ -0,0 +1,115 @@
+      - cp -r uppsrc /app/src
+      - cp .clang-format /app/src/uppsrc/.clang-format
+      - cp run-ide-install-host-deps.sh /app/src/run-ide-install-host-deps.sh
+      - cp -r /usr/lib/sdk/llvm17/lib /app
```
why do you copy this?

klugier replied:
We are using libclang for auto complete. We load libclang.so during runtime. For more info please look at this line https://github.com/ultimatepp/ultimatepp/blob/ef6d9db3de5db17b0ed3d5c55bf9d6812fe7f1eb/uppsrc/ide/clang/libclang.cpp#L30.

Without this line the auto complete feature (we call it assist++) doesn't work in flatpak sandbox environment.

hfiguiere commented on org.ultimatepp.TheIDE.yml:
```diff
@@ -0,0 +1,115 @@
+      - git apply flatpak.patch
+      - make -f Makefile -j $FLATPAK_BUILDER_N_JOBS
+      - mv ide theide
+      - install -D theide /app/bin/theide
```
```suggestion
      - install -Dm755 theide /app/bin/theide
```

klugier commented:
bot, build org.ultimatepp.TheIDE

bbhtt commented on org.ultimatepp.TheIDE.metainfo.xml:
```diff
@@ -0,0 +1,77 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<component type="desktop">
+    <id>org.ultimatepp.theide</id>
+    <launchable type="desktop-id">org.ultimatepp.TheIDE.desktop</launchable>
```
launchable doesn't match id

bbhtt commented on org.ultimatepp.TheIDE.yml:
```diff
@@ -0,0 +1,113 @@
+app-id: org.ultimatepp.TheIDE
```
```suggestion
id: org.ultimatepp.TheIDE
```

bbhtt commented on run-ide-install-host-deps.sh:
```diff
@@ -0,0 +1,73 @@
+}
+
+echo "You are running flatpaked TheIDE for the first time!"
+echo "It is highly recommended that you install all required dependencies on your host operating system."
```
Why do you need dependencies on host? You can make sdk the runtime or launch the flatpak with --devel.

klugier replied:
Hello @bbhtt, these dependencies are need to compile U++ applications on host machine directly from TheIDE (sandboxed app). Without them there will be compilation errors.

--devel is not an option here. We do not want to debug TheIDE itself. We want to compile completely new apps from the sandbox app.

hfiguiere commented on org.ultimatepp.TheIDE.yml:
```diff
@@ -0,0 +1,113 @@
+        only-arches: [x86_64]
+        dest-filename: host-spawn
+      - type: file
+        url: https://github.com/1player/host-spawn/releases/download/1.5.0/host-spawn-aarch64
```
can't these be built from source?

klugier replied:
Hello @hfiguiere,

The above statement downloads a static executable written in Go that works on a given operating system and architecture. So, there are no real benefits to building from scratch. Also, the author of the package provides pre-built binaries for Linux, x86_64, and arrch64 architectures. Which are the only architectures supported by Flathub at the momenet.

Currently, on the Flathub, there are two approaches to using host-spawn:
- https://github.com/flathub/com.visualstudio.code/blob/c2704e2c9c31ba12f8266fc27d5c73f6e68a51f4/com.visualstudio.code.yaml#L138 (VS Code - Downloading static executable)
- https://github.com/flathub/org.kde.qmlkonsole/blob/fe725223c81bd050da8b4dc96c10e17e8214a454/org.kde.qmlkonsole.json#L21 (QML Konsole - Downloading source package and building from scratch)

Personally, I think the first option is much more straightforward and relatively easy to maintain. You do not need to manually add all GoLang dependencies on your own each time you want to upgrade a package.

Please let me know what your recommendation is here. I can rework to build from sources if it is needed. However, I would like to avoid this in this particular case when everything that is needed is provided directly in a binary package. Also, looking at the current application distributed via Flathub, there is no unified way to do it.

klugier replied:
Hello @hfiguiere,

I manage to build host-spawn from sources. It was not straightforward, but in the end it was possible. I made decision to patch host-spawn, because it was using golang.org/x/sys in the unknown version (commit). So, I decided to change it to v0.15.0. The patch is under review in the upstream (https://github.com/1player/host-spawn/pull/30). Please keep in mind that using version from commit might be treated as security vulnerability, so it must be avoided whenever possible. I hope author will soon accept my PR and release new version. As far as I remember it was also the Flathub recommendation to build only from tagged releases.

Klugier

hfiguiere commented on org.ultimatepp.TheIDE.yml:
```diff
@@ -0,0 +1,113 @@
+        sha256: 23793a363a140139728a30cc050c35287d6ff97ee03125d62f8ad8acf06e40af
+        dest: 3p/uppsrc
+      - type: archive
+        url: https://github.com/ultimatepp/ultimatepp/releases/download/2023.2/umk-17045-linux-x86-64.tar.gz
```
can't this be built from source?

klugier replied:
This is probably doable. I will need to do investigation and I will back to you.

klugier replied:
Done!

klugier commented:
Hello @hfiguiere and @bbhtt,

I would like to ask if there is a possibility to make an exception to using "--talk-name=org.freedesktop.Flatpak" for this particular Flatpak. Without this exception, it will be very difficult to make a fully functional flatpak of TheIDE, an integrated development environment for the U++ framework. The idea behind this is to run all external binaries needed by TheIDE to compile, build, and debug projects on the host operating system instead of Sandbox. The host operating system can be easily expanded by packages that initially might not be needed by the user. For example, we offer a source package dedicated to MySQL, and it will only compile fine when an appropriate package is installed with the required ".so" and ".h" files. As far as I understand, users cannot install external packages or extensions to an already-installed Flatpak application. The other remarks raised in this PR seem reasonable to me, and they should be easily fixed and addressed. However, my priority here is to have a green light and a clear situation in the context of "--talk-name=org.freedesktop.Flatpak" permission.

Moreover, other applications in the same category (Integrated Development Environments) or similar categories hosted on FlatHub are using this permission:
- https://github.com/flathub/com.visualstudio.code/blob/c2704e2c9c31ba12f8266fc27d5c73f6e68a51f4/com.visualstudio.code.yaml#L29 (Visual Studio Code - downloaded over 2 million times)
- https://github.com/flathub/org.vim.Vim/blob/ea62568b727752a0c9bbc9564bf30623553666ea/org.vim.Vim.json#L14 (VIM - tool for developmenet)
- https://github.com/flathub/com.jetbrains.IntelliJ-IDEA-Community/blob/083b5292e216718f8655bf727f0d91ecd8b703c6/com.jetbrains.IntelliJ-IDEA-Community.yaml#L19 (InteliJ - famous IDE from JetBrains)
- https://github.com/flathub/org.godotengine.Godot/blob/e64401ff8dbbff755ff9e295a00cabf25c0f81db/org.godotengine.Godot.yaml#L49 (Godot - game engine - similar in many cases to TheIDE and U++ framework)
- much more...
So, it looks like the above projects manage to receive an exception for using "--talk-name=org.freedesktop.Flatpak" permission somehow.

I am open for a conversation with you when I will explain and tell you more about the IDE and the U++ framework to give you a better understanding. Please let me know if we can have a quick chat outside of GitHub.

I also encourage you to visit our official site and read Wikipedia article about the technology:
- https://www.ultimatepp.org/
- https://pl.wikipedia.org/wiki/Ultimate%2B%2B 
We might not be as recognizable as other tools, such as VS Code, but we have our ambitions. The Flatpak package on FlatHub will be an important step in fulfilling them. Also, we have a relatively large user base. Right now, we have around 680 stars on GitHub, and U++ has been downloaded more than 300,000 times from SourceForge.

Regards,
Zbigniew Rębacz

bbhtt commented:
> I would like to ask if there is a possibility to make an exception to using "--talk-name=org.freedesktop.Flatpak" 

Make a PR, please see the docs https://docs.flathub.org/docs/for-app-authors/linter#exceptions

klugier commented:
Thank you @bbhtt for understanding! I created PR https://github.com/flathub/flatpak-builder-lint/pull/227 as suggested.

klugier commented on org.ultimatepp.TheIDE.metainfo.xml:
```diff
@@ -0,0 +1,77 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<component type="desktop">
+    <id>org.ultimatepp.TheIDE</id>
+    <launchable type="desktop-id">org.ultimatepp.TheIDE.desktop</launchable>
```
@bbhtt and @hfiguiere is launchable OK or it should be changed? If it should be changed please let me know to what?

bbhtt replied:
looks correct now

klugier commented on org.ultimatepp.TheIDE.yml:
```diff
@@ -0,0 +1,114 @@
+      - mv 3p/Terminal/Terminal app/Terminal
+      - /app/bin/umk app/,3p/uppsrc UppTerm /app/bm/GCC.bm -brv +GUI,SHARED upp-term
+      - install -Dm755 upp-term /app/bin/upp-term
+      - rm -f /app/bin/umk
+      - rm -rf /app/bm/
```
It looks like when I put those into cleanup step, it doesn't work. After compiling UppTerm we do not longer need those files anymore in flatpak.

klugier commented:
Hello @hfiguiere and @bbhtt,

I addressed all the issues you raised during the code review, including those related to not building certain binaries from sources. Now, everything is built that way, including host-spawn and umk. Since everything is addressed, I would like to move forward with this PR. The first thing you should do is unlock the build by accepting https://github.com/flathub/flatpak-builder-lint/pull/227. After that, I will schedule the build and check it whenever it works. If it builds without trouble, I will test it. After successful tests, I think this PR should be merged with org.ultimatepp. The IDE package should be available on Flathub.

Also, please make the second review pass for this PR. I assume there might still be things that are worth optimizing in the context of being compatible with the latest Flathub standards.

Regards
Zbigniew Rębacz

klugier commented on patches/host-spawn.patch:
```diff
@@ -0,0 +1,24 @@
+diff --git a/go.mod b/go.mod
```
Security vulnerability patch. Waiting for owner review - https://github.com/1player/host-spawn/pull/30.

klugier commented on patches/icon.patch:
```diff
@@ -0,0 +1,246 @@
+diff --git a/uppsrc/ide/icon.svg b/uppsrc/ide/icon.svg
```
Already  in the upstream https://github.com/ultimatepp/ultimatepp/commit/123f8e0354e834101b094f486f7a53c465cc4d7c

klugier commented on patches/theide.patch:
```diff
@@ -0,0 +1,403 @@
+diff --git a/uppsrc/Core/config.h b/uppsrc/Core/config.h
```
https://github.com/ultimatepp/ultimatepp/pull/177 - will be merged to master branch once org.ultimatepp.TheIDE will be available on Flathub.

bbhtt commented on org.ultimatepp.TheIDE.yml:
```diff
@@ -0,0 +1,134 @@
+runtime: org.freedesktop.Platform
+runtime-version: "23.08"
+sdk: org.freedesktop.Sdk
+sdk-extensions:
+  [
+    "org.freedesktop.Sdk.Extension.llvm17",
+    "org.freedesktop.Sdk.Extension.golang",
+  ]
```
```suggestion
sdk-extensions:
  - org.freedesktop.Sdk.Extension.llvm17
  - org.freedesktop.Sdk.Extension.golang
```

bbhtt commented on org.ultimatepp.TheIDE.yml:
```diff
@@ -0,0 +1,134 @@
+      - install -D uppsrc/ide/icon.svg /app/share/icons/hicolor/scalable/apps/${FLATPAK_ID}.svg
+      - install -D uppsrc/ide/theide.desktop /app/share/applications/${FLATPAK_ID}.desktop
+      - desktop-file-edit --set-key=Icon --set-value=${FLATPAK_ID} /app/share/applications/${FLATPAK_ID}.desktop
+      - mkdir -p /app/src
+      - mkdir -p /app/lib
+      - mkdir -p /app/bm
```
```suggestion
      - mkdir -p /app/{src,lib,bm}
```

bbhtt commented on org.ultimatepp.TheIDE.yml:
```diff
@@ -0,0 +1,134 @@
+      - mv ide theide
+      - install -Dm755 theide /app/bin/theide
+      - install -Dm755 umk /app/bin/umk
+      - install -D uppsrc/ide/icon.svg /app/share/icons/hicolor/scalable/apps/${FLATPAK_ID}.svg
+      - install -D uppsrc/ide/theide.desktop /app/share/applications/${FLATPAK_ID}.desktop
```
```suggestion
      - install -Dm644 uppsrc/ide/icon.svg /app/share/icons/hicolor/scalable/apps/${FLATPAK_ID}.svg
      - install -Dm644 uppsrc/ide/theide.desktop /app/share/applications/${FLATPAK_ID}.desktop
```

bbhtt commented on org.ultimatepp.TheIDE.yml:
```diff
@@ -0,0 +1,134 @@
+    build-commands:
+      - install -Dm755 run-ide.sh /app/bin/run-ide.sh
+      - desktop-file-edit --set-key=Exec --set-value=run-ide.sh /app/share/applications/${FLATPAK_ID}.desktop
+      - install -D org.ultimatepp.TheIDE.metainfo.xml /app/share/metainfo/${FLATPAK_ID}.metainfo.xml
```
```suggestion
      - install -Dm644 org.ultimatepp.TheIDE.metainfo.xml /app/share/metainfo/${FLATPAK_ID}.metainfo.xml
```

bbhtt commented on org.ultimatepp.TheIDE.yml:
```diff
@@ -0,0 +1,134 @@
+        path: host-spawn/modules.txt
+        dest: vendor/
+
+  - name: final-cleanup
+    buildsystem: simple
+    build-commands:
+      - rm -f /app/bin/umk
+      - rm -rf /app/bm/
```
You can add it to cleanups: `/bin/umk` and `/bm`.

klugier commented:
FYI @bbhtt, Everything fixed in accordance to your remarks.

klugier commented:
Hello @bbhtt and @hfiguiere,

I have one more question. Is it possible to run a successful build for this PR without having https://github.com/flathub/flatpak-builder-lint/pull/227 merged? At the moment, I do not fully understand the procedure, and whenever it is possible to merge https://github.com/flathub/flatpak-builder-lint/pull/227 before merging this PR (chicken and egg problem).

Klugier

bbhtt commented:
> Is it possible to run a successful build for this PR without having

Nope, but you can build/test it locally and run all the checks https://docs.flathub.org/docs/for-app-authors/submission/#before-submission, if that works then it should build on Flathub too.

klugier commented:
Hello @bbhtt,

Everything builds fine, and I didn't find any issues during testing. All IDE functionalities work as expected. The only problem I see here is the output returned by the liner:
```
{
    "errors": [
        "appstream-screenshots-not-mirrored-in-ostree",
        "finish-args-flatpak-spawn-access"
    ]
}
```
The first error is obvious and expected. However, the second one, "appstream-screenshots-not-mirrored-in-ostree" is something new. I assume this error will not be present during the regular Flathub pipeline because screenshots mirroring to https://dl.flathub.org/repo/screenshots will be executed. Please let me know if I thought correctly. Alternatively, if it is something that requires action, please give me guidelines on how to fix it.

Also, it looks like this repository is stable, polished, and in accordance with guidelines. Can you tell me what the next steps will be and when I can expect the publication of this flatpak?

Klugier

bbhtt commented:
Both errors are fine, see the note after step #3 https://docs.flathub.org/docs/for-app-authors/submission/#before-submission

barthalion commented:
bot, build org.ultimatepp.TheIDE

barthalion commented:
bot, build org.ultimatepp.TheIDE

barthalion commented:
/merge
